File: src/services/spatial.ts
import { AWS_STATIONS } from '../data/stations';
import { SpatialNeighborInfo, WeatherObservation } from '../types';

/**
 * Calculates Great Circle distance between two lat/lng coordinates in kilometers using Haversine formula.
 */
export function haversineDistanceKm(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const R = 6371; // Earth's mean radius in km
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return Math.round(R * c * 10) / 10;
}

/**
 * Retrieves the k-nearest spatial neighbors for a target station within maxRadiusKm.
 */
export function getNearestNeighbors(
  targetStationId: string,
  maxNeighbors: number = 4,
  maxRadiusKm: number = 550
): { stationId: string; name: string; distanceKm: number }[] {
  const target = AWS_STATIONS.find((s) => s.id === targetStationId);
  if (!target) return [];

  return AWS_STATIONS
    .filter((s) => s.id !== targetStationId)
    .map((s) => ({
      stationId: s.id,
      name: s.name,
      distanceKm: haversineDistanceKm(target.lat, target.lng, s.lat, s.lng)
    }))
    .filter((s) => s.distanceKm <= maxRadiusKm)
    .sort((a, b) => a.distanceKm - b.distanceKm)
    .slice(0, maxNeighbors);
}

/**
 * Performs cross-station spatial consistency check against neighbor current readings.
 */
export function analyzeSpatialConsensus(
  targetStationId: string,
  currentObs: WeatherObservation,
  allLatestObservations: Map<string, WeatherObservation>
): {
  neighbors: SpatialNeighborInfo[];
  spatialScore: number; // 0 (perfect agreement) to 100 (severe spatial disagreement)
  isConsensusEvent: boolean; // True if neighbors are also observing the sudden shift (Weather event)
  medianTemp: number;
  medianPressure: number;
  medianHumidity: number;
  flags: string[];
} {
  const neighborsList = getNearestNeighbors(targetStationId, 4, 600);
  const flags: string[] = [];

  if (neighborsList.length === 0) {
    return {
      neighbors: [],
      spatialScore: 0,
      isConsensusEvent: false,
      medianTemp: currentObs.temperature,
      medianPressure: currentObs.pressure,
      medianHumidity: currentObs.humidity,
      flags: ['No spatial neighbours within 600km radius']
    };
  }

  const validNeighbors: SpatialNeighborInfo[] = [];
  const neighborTemps: number[] = [];
  const neighborPressures: number[] = [];
  const neighborHumidities: number[] = [];

  for (const n of neighborsList) {
    const nObs = allLatestObservations.get(n.stationId);
    if (nObs && !nObs.isMissing) {
      const tempDelta = Math.round((currentObs.temperature - nObs.temperature) * 10) / 10;
      validNeighbors.push({
        stationId: n.stationId,
        name: n.name,
        distanceKm: n.distanceKm,
        temperature: nObs.temperature,
        pressure: nObs.pressure,
        humidity: nObs.humidity,
        tempDelta,
        isCorrelated: Math.abs(tempDelta) < 5.0
      });
      neighborTemps.push(nObs.temperature);
      neighborPressures.push(nObs.pressure);
      neighborHumidities.push(nObs.humidity);
    }
  }

  if (validNeighbors.length === 0) {
    return {
      neighbors: [],
      spatialScore: 0,
      isConsensusEvent: false,
      medianTemp: currentObs.temperature,
      medianPressure: currentObs.pressure,
      medianHumidity: currentObs.humidity,
      flags: ['Neighbour telemetry currently offline']
    };
  }

  const median = (arr: number[]) => {
    const s = [...arr].sort((a, b) => a - b);
    const mid = Math.floor(s.length / 2);
    return s.length % 2 !== 0 ? s[mid] : (s[mid - 1] + s[mid]) / 2;
  };

  const medianTemp = median(neighborTemps);
  const medianPressure = median(neighborPressures);
  const medianHumidity = median(neighborHumidities);

  const deltaTemp = Math.abs(currentObs.temperature - medianTemp);
  const deltaPressure = Math.abs(currentObs.pressure - medianPressure);
  const deltaHumidity = Math.abs(currentObs.humidity - medianHumidity);

  let spatialScore = 0;

  if (deltaTemp > 7.0) {
    spatialScore += Math.min(50, (deltaTemp - 7.0) * 10);
    flags.push(`Temperature deviates by ${deltaTemp.toFixed(1)}°C from spatial cluster median (${medianTemp.toFixed(1)}°C)`);
  }

  if (deltaPressure > 15.0) {
    spatialScore += Math.min(30, (deltaPressure - 15.0) * 4);
    flags.push(`Pressure deviates by ${deltaPressure.toFixed(1)} hPa from cluster median (${medianPressure.toFixed(1)} hPa)`);
  }

  if (deltaHumidity > 35) {
    spatialScore += Math.min(20, (deltaHumidity - 35) * 1.5);
    flags.push(`Humidity deviates by ${deltaHumidity.toFixed(0)}% from cluster median (${medianHumidity.toFixed(0)}%)`);
  }

  // Determine if neighbors also moved significantly together (Weather Event vs Isolated Sensor Fault)
  // Calculate variance among neighbors themselves
  let neighborTempVar = 0;
  if (neighborTemps.length > 1) {
    const mean = neighborTemps.reduce((a, b) => a + b, 0) / neighborTemps.length;
    neighborTempVar = neighborTemps.reduce((acc, val) => acc + Math.pow(val - mean, 2), 0) / neighborTemps.length;
  }

  // If neighbors also have higher variation or moved together in a similar gradient, it suggests a dynamic frontal boundary
  const isConsensusEvent = neighborTempVar > 12.0 || (validNeighbors.filter(n => Math.abs(n.tempDelta) < 3.5).length >= 2 && spatialScore < 45);

  return {
    neighbors: validNeighbors,
    spatialScore: Math.min(100, Math.round(spatialScore)),
    isConsensusEvent,
    medianTemp,
    medianPressure,
    medianHumidity,
    flags
  };
}

File: backend/app/detection/spatial.py
import math
import numpy as np
from typing import Dict, Any, List, Tuple
from backend.app.simulator.network import STATIONS_METADATA
from backend.app.config import settings

def haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Calculate the great circle distance between two points on the earth in km."""
    r = 6371.0  # Earth's radius in kilometers
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = math.sin(dlat / 2.0)**2 + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(dlon / 2.0)**2
    c = 2.0 * math.atan2(math.sqrt(a), math.sqrt(1.0 - a))
    return r * c

class SpatialCrossStationEngine:
    def __init__(self, k_neighbours: int = 5, max_radius_km: float = 450.0):
        self.k = k_neighbours
        self.max_radius_km = max_radius_km
        self.station_coords = {s["station_id"]: (s["lat"], s["lon"]) for s in STATIONS_METADATA}
        self.neighbour_cache: Dict[str, List[Tuple[str, float]]] = {}
        self._build_neighbour_graph()

    def _build_neighbour_graph(self):
        """Precompute distance rankings for all stations."""
        for s1 in STATIONS_METADATA:
            sid1 = s1["station_id"]
            lat1, lon1 = s1["lat"], s1["lon"]
            distances = []
            for s2 in STATIONS_METADATA:
                sid2 = s2["station_id"]
                if sid1 == sid2:
                    continue
                lat2, lon2 = s2["lat"], s2["lon"]
                dist = haversine_km(lat1, lon1, lat2, lon2)
                distances.append((sid2, dist))
            distances.sort(key=lambda x: x[1])
            self.neighbour_cache[sid1] = distances

    def get_neighbours(self, station_id: str) -> List[Tuple[str, float]]:
        all_dists = self.neighbour_cache.get(station_id, [])
        k = getattr(settings, 'SPATIAL_NEIGHBOUR_COUNT', self.k)
        max_radius = getattr(settings, 'SPATIAL_MAX_RADIUS_KM', self.max_radius_km)
        # Filter within max radius and take top k
        close = [pair for pair in all_dists if pair[1] <= max_radius]
        if len(close) < k and all_dists:
            return all_dists[:k]
        return close[:k]

    def analyze(
        self,
        station_id: str,
        current_observation: Dict[str, Any],
        all_observations: Dict[str, Dict[str, Any]]
    ) -> Tuple[float, str, List[str], Dict[str, Any]]:
        """
        Returns:
            spatial_anomaly_score (0-100),
            spatial_classification ("NORMAL", "ISOLATED_SENSOR_FAULT", "GENUINE_WEATHER_EVENT", "INSUFFICIENT_DATA"),
            evidence (List[str]),
            spatial_metrics (Dict)
        """
        evidence = []
        spatial_metrics = {}

        t = current_observation.get("temperature_c")
        p = current_observation.get("pressure_hpa")
        rh = current_observation.get("relative_humidity")

        if t is None:
            return 0.0, "NORMAL", [], {}

        neighbours = self.get_neighbours(station_id)
        if not neighbours:
            return 0.0, "INSUFFICIENT_DATA", ["No adjacent monitoring stations within spatial correlation radius."], {}

        neighbour_obs = []
        neighbour_distances = []
        for nid, dist in neighbours:
            if nid in all_observations:
                n_data = all_observations[nid]
                if n_data.get("temperature_c") is not None:
                    neighbour_obs.append(n_data)
                    neighbour_distances.append((nid, dist, n_data["temperature_c"]))

        if len(neighbour_obs) < 2:
            return 0.0, "INSUFFICIENT_DATA", ["Insufficient active neighbour telemetry for spatial cross-validation."], {}

        # Extract neighbour values
        n_temps = [o["temperature_c"] for o in neighbour_obs]
        n_pressures = [o["pressure_hpa"] for o in neighbour_obs if o.get("pressure_hpa") is not None]
        n_rhs = [o["relative_humidity"] for o in neighbour_obs if o.get("relative_humidity") is not None]

        # Calculate neighbour central tendencies
        n_mean_t = float(np.mean(n_temps))
        n_med_t = float(np.median(n_temps))
        n_std_t = float(np.std(n_temps))

        spatial_metrics["neighbour_count"] = len(neighbour_obs)
        spatial_metrics["neighbour_mean_t"] = round(n_mean_t, 2)
        spatial_metrics["neighbour_median_t"] = round(n_med_t, 2)
        spatial_metrics["neighbour_std_t"] = round(n_std_t, 2)
        spatial_metrics["neighbour_details"] = [
            {"station_id": nid, "distance_km": round(dist, 1), "temperature_c": temp}
            for nid, dist, temp in neighbour_distances
        ]

        # Target station deviation from neighbour mean
        dev_t = t - n_mean_t
        spatial_metrics["deviation_t"] = round(dev_t, 2)

        # Standardized spatial Z-score
        z_spatial_t = dev_t / max(n_std_t, 0.4)
        spatial_metrics["z_spatial_t"] = round(z_spatial_t, 2)

        score = 0.0
        classification = "NORMAL"

        # Check pressure and humidity spatial alignment if available
        dev_p = 0.0
        if p is not None and n_pressures:
            n_mean_p = float(np.mean(n_pressures))
            dev_p = p - n_mean_p
            spatial_metrics["neighbour_mean_p"] = round(n_mean_p, 2)
            spatial_metrics["deviation_p"] = round(dev_p, 2)

        # Core Spatial Intelligence:
        # 1. ISOLATED SENSOR FAULT:
        # Target station deviates sharply (|dev_t| > 7C or |z| > 3.0), but neighbours are in close agreement (n_std_t < 2.5C)
        if abs(dev_t) > 6.0 and abs(z_spatial_t) > 2.8:
            if n_std_t < 3.2:
                # Neighbours are tight and calm, this station is an outlier
                score = min(98.0, 50.0 + abs(dev_t) * 4.0)
                classification = "ISOLATED_SENSOR_FAULT"
                evidence.append(
                    f"Probable Sensor Fault — Observation ({t:.1f}°C) deviates by {dev_t:+.1f}°C ({abs(z_spatial_t):.1f}σ) from {len(neighbour_obs)} nearby stations (mean: {n_mean_t:.1f}°C, σ: {n_std_t:.1f}°C) which remain normal."
                )
            else:
                # Neighbours themselves have high spread, could be a weather boundary / front
                score = min(60.0, 30.0 + abs(dev_t) * 2.0)
                classification = "SUSPICIOUS_FRONT"
                evidence.append(
                    f"Spatial gradient detected: Station deviates by {dev_t:+.1f}°C from regional mean, with high neighbour variability (σ={n_std_t:.1f}°C)."
                )

        # 2. GENUINE WEATHER EVENT:
        # If target station experienced a large shift (e.g., from its own historical baseline)
        # but the neighbour mean also experienced a large consistent shift or matches the target station closely:
        # Or if multiple nearby stations report synchronized abnormal values:
        if abs(dev_t) <= 3.5:
            # Station is consistent with its neighbours
            score = max(0.0, abs(dev_t) * 2.0)
            classification = "NORMAL"
        
        # Detect if a regional weather event is in progress across the cluster
        # Check if the neighbour cluster as a whole shifted significantly from regional baseline
        meta = next((s for s in STATIONS_METADATA if s["station_id"] == station_id), None)
        if meta:
            base_t = meta["base_t"]
            cluster_shift = n_mean_t - base_t
            station_shift = t - base_t
            # If both station and neighbours shifted by > 4.5C in the SAME direction
            if abs(station_shift) > 4.5 and abs(cluster_shift) > 4.0 and (station_shift * cluster_shift > 0):
                classification = "GENUINE_WEATHER_EVENT"
                score = 30.0  # Moderate anomaly score for weather event, but NOT sensor fault!
                evidence.append(
                    f"Genuine Weather Event — Multiple nearby stations show consistent changes (cluster shift: {cluster_shift:+.1f}°C, target shift: {station_shift:+.1f}°C). Spatial consistency confirmed."
                )

        spatial_metrics["score"] = round(score, 1)
        spatial_metrics["classification"] = classification
        return score, classification, evidence, spatial_metrics
